nbodykit/plugins/datasource/Zheng07HOD.py

```diff
@@ -0,0 +1,177 @@
+from nbodykit.extensionpoints import DataSource, datasources
+
+import numpy
+import logging
+
+logger = logging.getLogger('Zheng07HOD')
+
+class Zheng07HODDataSource(DataSource):
+    """
+    A `DataSource` that uses the HOD prescription of 
+    Zheng et al. 2007 to populate an input halo catalog with galaxies, 
+    and returns the (Position, Velocity) of those galaxies
+    
+    The mock population is done using `halotools` (http://halotools.readthedocs.org)
+    The HOD model is of the commonly-used form:
+    
+    * logMmin: Minimum mass required for a halo to host a central galaxy
+    * sigma_logM: Rate of transition from <Ncen>=0 --> <Ncen>=1
+    * alpha: Power law slope of the relation between halo mass and <Nsat>
+    * logM0: Low-mass cutoff in <Nsat>
+    * logM1: Characteristic halo mass where <Nsat> begins to assume a power law form
+    
+    See the documentation for the `halotools` builtin Zheng07 HOD model, 
+    for further details regarding the HOD 
+    """
+    plugin_name = "Zheng07HOD"
+    
+    def __init__(self, halocat, redshift, logMmin=13.031, sigma_logM=0.38, 
+                    alpha=0.76, logM0=13.27, logM1=14.08, rsd=None):
+        """
+        Default values for HOD values from Reid et al. 2014
+        """
+        # load halotools
+        try:
+            from halotools import sim_manager
+        except:
+            name = self.__class__.__name__
+            raise ValueError("`halotools` must be installed to use '%s' DataSource" %name)
+            
+        # need cosmology
+        if self.cosmo is None:
+            raise AttributeError("a cosmology instance is required to populate an HOD")
+        
+        # set defaults and load the rest
+        sim_manager.sim_defaults.default_cosmology = self.cosmo.engine
+        sim_manager.sim_defaults.default_redshift = self.redshift
+        from halotools import empirical_models as em_models
+
+        # explicitly set an analytic mass-concentration relation
+        sats_prof_model = em_models.NFWPhaseSpace(conc_mass_model='dutton_maccio14')
+        
+        # build the model and set defaults
+        base_model = em_models.PrebuiltHodModelFactory('zheng07')
+        self.model = em_models.HodModelFactory(baseline_model_instance=base_model, 
+                                                satellites_profile=sats_prof_model)
+        for param in self.model.param_dict:
+            self.model.param_dict[param] = getattr(self, param)
+        
+        # data from the halo catalog
+        [[Position, Velocity, Mass]] = self.halocat.read(['Position','Velocity', 'Mass'], full=True)
```

Comment:
Ok, that's clarifying. With Scenario 1, there are several things that you need to be careful about. For example, how is Lbox being set for the subvolumes? Is it the size of the partition, or is it the periodic boundary conditions of the full snapshot? This is important because if the `enforce_periodic_boundary_conditions` function detects points (satellites) beyond Lbox, their positions are wrapped back inside box. Scenario 2 does not have this problem and is generally simpler as far as Halotools is concerned, but I understand why you'd rather not suffer the runtime hit of recollecting and rescattering (or is this not a worry?)

As for correlation of random number generators, I have not had to worry about this before now so this is untested. I'd be discuss how to experiment with this for your use case though.